### PR TITLE
Include new UART library with buffering.

### PR DESCRIPTION
**Pull request by ihartwig**
* add Gock/Fleury UART library with support for most AVR chips
* add extra calls to use library with avr-libc file streams
* update radio_buggy_mega project to use new library with 2 uarts

TODO:
* need to add some docs
* #defines do not work across the library/program boundary well. options?

### CODE REVIEW

**Comment by ihartwig** [comment]
We should test this to see if it works besides compiling.

Tagging @ramensandwich, @TrevorDecker .

**Comment by ramensandwich** [comment]
Instead of using #defines, can we create some constant variables in comp.h?

**Comment by TrevorDecker** [comment]
For: #defines do not work across the library/program boundary well. options?
was this solved in todays meeting or is it still an open problem. If it is can you provide more detail as to why it does not work?  

If the problem is that #define is not a valid command on all source code then we could write a script which will replace specific strings with other specific strings aka make our own pre compiler.

**Comment by ihartwig** [comment]
This was partially answered while we were reviewing because I thought I could get gcc to optimize out functions that we weren't using (like in http://stackoverflow.com/questions/6687630/c-c-gcc-ld-remove-unused-symbols), but that doesn't work in practice. Removing all references to a function doesn't make the code size go down but disabling a UART by #define does.

Fortunately, we have a project configuration file called Makefile. See 2b9f09c.

**Comment by msebek** [comment]
Everything here looks sensible; is it ready to be merged in?

**Comment by TrevorDecker** [comment]
I am good with these changes but I think Ian has recently made some more changes to these files, removing a couple bugs. I would rather limit the number of changes to master, and keep master as stable as possible.  Ian can you submit a new pull request with the code that you deployed to the buggy on sunday.

**Comment by ihartwig** [comment]
No. I will not open it in another pull request so that we loose the conversation that exists here.

**Comment by ihartwig** [comment]
Small bug fixed in d65c074 where baud rate was provided in the wrong format to uartN_init(). This should be ready now if it looks good to you guys.
Updated branch in prep for merge.

File: real_time/arduino_src/radio_buggy_mega/Makefile [hunk 1]
```diff
@@ -65,11 +65,12 @@ TARGET = main
 
 # List C source files here. (C dependencies are automatically generated.)
 SRC = $(TARGET).c
-SRC += uart.c
+# SRC += uart.c
```
Comment (TrevorDecker) [inline real_time/arduino_src/radio_buggy_mega/Makefile]: is uart.c going to be something that we use in the future?

Comment (ihartwig) [inline real_time/arduino_src/radio_buggy_mega/Makefile]: Nope. I should probably roll in deleting this file.

**Pull request merged**